src/aggregator/aggregator/flush_mgr.go

```diff
@@ -309,17 +317,39 @@ func (mgr *flushManager) computeFlushIntervalOffset(flushInterval time.Duration)
 	// Otherwise the offset is determined based on jittering.
 	maxJitter := mgr.maxJitterFn(flushInterval)
 	jitterNanos := mgr.randFn(int64(maxJitter))
-	return time.Duration(jitterNanos)
+	value := time.Duration(jitterNanos)
+	return value
 }
 
 // NB(xichen): apparently timer.Reset() is more difficult to use than I originally
 // anticipated. For now I'm simply waking up every second to check for updates. Maybe
 // when I have more time I'll spend a few hours to get timer.Reset() right and switch
 // to a timer.Start + timer.Stop + timer.Reset + timer.After based approach.
 func (mgr *flushManager) flush() {
-	defer mgr.Done()
+	var tasks []flushTask
+	defer func() {
+		// First wait for all outstanding tasks to complete.
+		for _, t := range tasks {
+			t.WaitComplete()
+		}
+		// Mark as done.
+		mgr.Done()
+	}()
 
 	for {
+		// Remove any completed tasks tracking.
+		pending := tasks[:0]
+		for i, t := range tasks {
+			// Nil ref to task otherwise reused slice would still hold ref.
```

Comment:
Would it still hold the ref if you append over it, and then set `tasks := pending[:len(pending)]`?